### PR TITLE
Customize Discord OAuth provider

### PR DESCRIPTION
**Pull request by jchristgit**
This PR adds a new app, called `oauth`, and changes or adds the following on top of the previously used OAuth provider:

- removes the unused `email` scope
- adds the `guild` scope
- stores guild data in the `extra_data['guilds']` key (which is an array of [guild objects](https://discordapp.com/developers/docs/resources/guild#guild-object))
- adds a method to retrieve a user's avatar URL
- adds a method to open a direct message with a user ("profile URL")

### CODE REVIEW

**Comment by emmiegit** [comment]
Nice, more proper objects returned from django.
Not totally related, but is the weirdness with "first name" and "last name" fixable?

File: oauth/provider.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        corresponding default avatar from Discord instead.
+        """
+
+        if 'avatar' in self.account.extra_data:
+            avatar_hash = self.account.extra_data['avatar']
+            user_id = self.account.extra_data['id']
+            return f'https://cdn.discordapp.com/avatars/{user_id}/{avatar_hash}.png'
+
+        discriminator = self.account.extra_data['discriminator']
+        return f'https://cdn.discordapp.com/embed/avatars/{discriminator % 5}.png'
```
Comment (emmiegit) [inline oauth/provider.py:29]: TIL this is how default avatars work

File: oauth/views.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        headers = {
+            'Authorization': f'Bearer {token.token}',
+            'Content-Type': 'application/json'
+        }
+        profile_data = requests.get(self.profile_url, headers=headers)
+        guild_data = requests.get(self.guild_url, headers=headers)
+
+        extra_data = {
+            **profile_data.json(),
+            'guilds': guild_data.json()
```
Comment (emmiegit) [inline oauth/views.py]: I think kwargs should be placed after. I know this isn't exactly what this is, but that's how Python likes functions, so I think it makes sense here.

**Review by emmiegit — APPROVED** [review]

**Comment by emmiegit** [comment]
Btw why "`discoauth`"?

File: oauth/provider.py (hunk 1 above)
Comment (jchristgit) [reply oauth/provider.py:29]: Yup: https://discordapp.com/developers/docs/reference#image-formatting-cdn-endpoints

**Comment by jchristgit** [comment]
`discord` is already used by the built-in OAuth2 provider, but using a custom ID is desired. Initially I went with `custom_discord` or `discord_custom`, but that looks pretty horrible, so I used `discoauth`.

**Comment by jchristgit** [comment]
> Not totally related, but is the weirdness with "first name" and "last name" fixable?

It is, but after some digging it seems like it's only fixable through forking the repo again and I'd like to avoid that. We have all the necessary data from Discord to display the actual username and discriminator.

**Pull request merged**